File: interface/src/main/java/xsbti/OutputLocation.java
/* sbt -- Simple Build Tool
 * Copyright 2008, 2009 Mark Harrah
 */
package xsbti;

import java.io.File;
import java.io.IOException;

/**
 * Interface differentiate Jar outputs from Directory outputs.
 */
public abstract class OutputLocation {
  protected final File file;
  protected OutputLocation(File f) {
    this.file = f;
  }

  /**
   * The directory which this output will be placed into.
   */
  protected abstract File directory();

  /**
   * @return The File for this OutputLocation, which may be either a jar or directory.
   */
  public final File file() {
    return this.file;
  }

  /**
   * Prepares this location to receive output by creating any necessary parent directories.
   */
  public final void prepare() throws IOException {
    this.directory().mkdirs();
  }

  /**
   * @return Creates an OutputLocation for the given file, which will represent either
   * a jar or directory.
   */
  public static OutputLocation create(File file) {
    if (endsWithJar(file)) {
      return new JarOutputLocation(file);
    }
    return new DirectoryOutputLocation(file);
  }

  protected static boolean endsWithJar(File file) {
    return file.getName().toLowerCase().endsWith(".jar");
  }
}

File: compile/interface/src/main/scala/xsbt/OutputLocation.scala
/* sbt -- Simple Build Tool
 * Copyright 2008, 2009 Mark Harrah
 */
package xsbt

import java.io.File

/**
 * Trait to differentiate Jar outputs from Directory outputs.
 */
sealed trait OutputLocation {
  def file: File
  def prepare(): Unit
}
object OutputLocation {
  /**
   * @return Creates an OutputLocation for the given file, which will represent either
   * a jar or directory.
   */
  def apply(file: File) =
    if (endsWithJar(file)) {
      Jar(file)
    } else {
      Directory(file)
    }

  case class Jar(file: File) extends OutputLocation {
    assert(endsWithJar(file), s"${file} is not a jarfile.")
    def prepare() = IO.createDirectory(file.getParentFile)
  }
  case class Directory(file: File) extends OutputLocation {
    assert(!endsWithJar(file), s"${file} is (probably) not a directory.")
    def prepare() = IO.createDirectory(file)
  }

  private def endsWithJar(file: File): Boolean =
    file.getName.toLowerCase.endsWith(".jar")
}
